File: frontend/src/api/finance.ts
import api from "./client";

// ── Types ───────────────────────────────────────────────

export interface Supplier {
  id: string;
  name: string;
  rfc?: string;
  contact_name?: string;
  phone?: string;
  email?: string;
  payment_terms: number;
  notes?: string;
  is_active: boolean;
  created_at: string;
}

export interface Invoice {
  id: string;
  project_id?: string;
  invoice_number: string;
  cfdi_uuid?: string;
  client: string;
  issue_date: string;
  due_date: string;
  subtotal: number;
  tax_rate: number;
  tax_amount: number;
  total: number;
  status: string;
  payment_date?: string;
  pdf_url?: string;
  xml_url?: string;
  created_at: string;
}

export interface Expense {
  id: string;
  supplier_id?: string;
  project_id?: string;
  category: string;
  description: string;
  amount: number;
  tax_amount: number;
  total: number;
  date: string;
  due_date?: string;
  status: string;
  payment_date?: string;
  invoice_ref?: string;
  supplier_name?: string;
  created_at: string;
}

export interface AgingBucket {
  client: string;
  current: number;
  days_31_60: number;
  days_61_90: number;
  over_90: number;
  total: number;
}

export interface DashboardKPIs {
  machines: {
    total: number;
    by_status: Record<string, number>;
  };
  month_revenue: number;
  month_expenses: number;
  margin_percentage: number;
  active_projects: number;
}

export interface CashflowItem {
  month: string;
  income: number;
  expenses: number;
}

export interface ProfitabilityItem {
  project_code: string;
  project_name: string;
  contract_amount: number;
  revenue: number;
  expenses: number;
  profit: number;
  margin: number;
  progress: number;
}

// ── API ─────────────────────────────────────────────────

export const financeApi = {
  // Invoices
  getInvoices: (params?: {
    status?: string;
    client?: string;
    project_id?: string;
  }) => api.get<Invoice[]>("/invoices", { params }),

  getInvoice: (id: string) => api.get<Invoice>(`/invoices/${id}`),

  createInvoice: (data: Partial<Invoice>) =>
    api.post<Invoice>("/invoices", data),

  updateInvoice: (id: string, data: Partial<Invoice>) =>
    api.patch<Invoice>(`/invoices/${id}`, data),

  deleteInvoice: (id: string) => api.delete(`/invoices/${id}`),

  getReceivables: () => api.get<AgingBucket[]>("/invoices/receivables"),

  // Expenses
  getExpenses: (params?: {
    category?: string;
    status?: string;
    project_id?: string;
    supplier_id?: string;
  }) => api.get<Expense[]>("/expenses", { params }),

  getExpense: (id: string) => api.get<Expense>(`/expenses/${id}`),

  createExpense: (data: Partial<Expense>) =>
    api.post<Expense>("/expenses", data),

  updateExpense: (id: string, data: Partial<Expense>) =>
    api.patch<Expense>(`/expenses/${id}`, data),

  deleteExpense: (id: string) => api.delete(`/expenses/${id}`),

  // Suppliers
  getSuppliers: (search?: string) =>
    api.get<Supplier[]>("/expenses/suppliers", { params: { search } }),

  createSupplier: (data: Partial<Supplier>) =>
    api.post<Supplier>("/expenses/suppliers", data),

  updateSupplier: (id: string, data: Partial<Supplier>) =>
    api.patch<Supplier>(`/expenses/suppliers/${id}`, data),

  // Reports / Dashboard
  getDashboardKPIs: () => api.get<DashboardKPIs>("/reports/dashboard"),

  getCashflow: (months?: number) =>
    api.get<CashflowItem[]>("/reports/cashflow", { params: { months } }),

  getProfitability: () =>
    api.get<ProfitabilityItem[]>("/reports/profitability"),
};

File: backend/app/schemas/finance.py
"""Schemas Pydantic para finanzas: facturas, gastos, proveedores."""

from datetime import date, datetime
from decimal import Decimal
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field


# ── Supplier ──────────────────────────────────────────

class SupplierCreate(BaseModel):
    name: str = Field(min_length=2, max_length=255)
    rfc: Optional[str] = None
    contact_name: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[str] = None
    payment_terms: int = 30
    notes: Optional[str] = None


class SupplierResponse(BaseModel):
    id: UUID
    name: str
    rfc: Optional[str] = None
    contact_name: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[str] = None
    payment_terms: int
    notes: Optional[str] = None
    is_active: bool
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Invoice ───────────────────────────────────────────

class InvoiceCreate(BaseModel):
    project_id: Optional[UUID] = None
    invoice_number: str = Field(min_length=1, max_length=50)
    cfdi_uuid: Optional[str] = None
    client: str = Field(min_length=2, max_length=255)
    issue_date: date
    due_date: date
    subtotal: Decimal = Field(ge=0)
    tax_rate: Decimal = Decimal("0.16")
    status: str = "borrador"


class InvoiceUpdate(BaseModel):
    client: Optional[str] = None
    issue_date: Optional[date] = None
    due_date: Optional[date] = None
    subtotal: Optional[Decimal] = None
    tax_rate: Optional[Decimal] = None
    status: Optional[str] = None
    payment_date: Optional[date] = None
    cfdi_uuid: Optional[str] = None
    pdf_url: Optional[str] = None
    xml_url: Optional[str] = None


class InvoiceResponse(BaseModel):
    id: UUID
    project_id: Optional[UUID] = None
    invoice_number: str
    cfdi_uuid: Optional[str] = None
    client: str
    issue_date: date
    due_date: date
    subtotal: Decimal
    tax_rate: Decimal
    tax_amount: Decimal
    total: Decimal
    status: str
    payment_date: Optional[date] = None
    pdf_url: Optional[str] = None
    xml_url: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Expense ───────────────────────────────────────────

class ExpenseCreate(BaseModel):
    supplier_id: Optional[UUID] = None
    project_id: Optional[UUID] = None
    category: str
    description: str = Field(min_length=2, max_length=500)
    amount: Decimal = Field(gt=0)
    tax_amount: Decimal = Decimal("0")
    date: date
    due_date: Optional[date] = None
    status: str = "pendiente"
    invoice_ref: Optional[str] = None


class ExpenseUpdate(BaseModel):
    category: Optional[str] = None
    description: Optional[str] = None
    amount: Optional[Decimal] = None
    tax_amount: Optional[Decimal] = None
    date: Optional[date] = None
    due_date: Optional[date] = None
    status: Optional[str] = None
    payment_date: Optional[date] = None
    invoice_ref: Optional[str] = None


class ExpenseResponse(BaseModel):
    id: UUID
    supplier_id: Optional[UUID] = None
    project_id: Optional[UUID] = None
    category: str
    description: str
    amount: Decimal
    tax_amount: Decimal
    total: Decimal
    date: date
    due_date: Optional[date] = None
    status: str
    payment_date: Optional[date] = None
    invoice_ref: Optional[str] = None
    supplier_name: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Aging / Receivables ───────────────────────────────

class AgingBucket(BaseModel):
    client: str
    current: Decimal  # 0-30 días
    days_31_60: Decimal
    days_61_90: Decimal
    over_90: Decimal
    total: Decimal
